### ISSUE TITLE
Frameit ignoring dotenv. Documentation questions/comments.

### ISSUE THREAD

**Issue by timothycosta** [fastlane/fastlane#8673]
### New Issue Checklist

- [ X ] Updated fastlane to the latest version
- [ X ] I have read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)

### Issue Description

I have created a number of branded apps and separated their settings out into various `.env` files.  Our core app is in `.env` and the branded apps are in `.env.en`, `.env.de` etc... by language code.  App metadata/screenshots are in directories also named by language code, with the main app in `all`.  Directory structure like `my_project/fastlane/all/screenshots...` `my_project/fastlane/en/screenshots...`.

When running frameit, it seems like my .env files are ignored (or more correctly, is not configurable), and frameit does a brute force search of every directory for possible screenshots.  I only figured this out on a hunch after I changed the directory in my main `.env` file to something else and frameit kept working on the previous directory which was no longer referenced anywhere except for the filesystem.  At that point frameit even tried to frame my `background.png` which I had moved directly into `fastlane/` so that it could be referenced by all my projects.

It seems like frameit works significantly differently from the rest of fastlane.  For example, `Framefile.json` is required to reside within the `screenshots` directory, rather than having a ruby `Framefile` where you can define a path to `Framefile.json`.  In our case a separate `Framefile` would be very beneficial since I currently will need to copy `Framefile.json` into 14 different `screenshots` folders and keep them in sync.

### Questions:
1. Is there currently any way to get frameit to work on one `[app, directory, .env]` at a time?
2. Frameit doesn't appear to delete the unframed files.  How does `deliver` determine which files are the correct ones?  Does it compare filenames and prioritize those ending in `_framed` or just upload everything?
3. Can a file be excluded from framing?

### Suggestions:
1. The current behavior should be added to the documentation regarding how frameit determines which files/directories to work on/in.
1. Allow for a separate `Framefile` to match other fastlane tools.  Could include something like `output_directory` which matches `Snapfile`, `keywords_directory`, `titles_directory`, `fonts_directory`, `json_path`...
2. Documentation for `Framefile.json` could be improved.  For example, it's not clear how the filters apply.  Do they get used only once, or for every matching occurrence of that string in the filename?  Probably the latter, but being explicit is nice.  What does `padding` do?  Is it horizontal, vertical?  Above and below the text or just below?
3. Documentation should be provided regarding the `background` file.  It appears this is used for all images, but there is no suggestion on size, or mention of how it gets scaled.  Relevant if you wanted to use a gradient image.  (Also, why is a file required as opposed to just a color string in `Framefile.json`?)

##### Complete output when running fastlane, including the stack trace and command used

I'm including two "outputs".  Our main app is contained within the `all` directory which by chance starts with `a` and so it gets grabbed first in output 1.  Then I renamed `all` to `zall` and frameit grabbed `de` first and started framing those files.  In both cases I cancelled out.

##### Output 1

Timothys-Mac-Mini:iLingQ timothycosta$ fastlane frames
[10:14:46]: -------------------------------------------------
[10:14:46]: --- Step: Verifying required fastlane version ---
[10:14:46]: -------------------------------------------------
[10:14:46]: Your fastlane version 2.23.0 matches the minimum requirement of 2.20.0  ✅
[10:14:46]: ------------------------------
[10:14:46]: --- Step: default_platform ---
[10:14:46]: ------------------------------
[10:14:46]: Driving the lane 'ios frames' 🚀
[10:14:46]: ---------------------
[10:14:46]: --- Step: frameit ---
[10:14:46]: ---------------------
[10:14:46]: Framing screenshots at path ./fastlane/
[10:14:46]: Using device frames version 'latest'
[10:14:46]: Framing screenshot './all/screenshots/en-US/1_ipadPro_1.screenshot-ipad-landscape-1 (1).png'
[10:14:52]: Added frame: '/Users/timothycosta/Desktop/lingq/iLingQ/fastlane/all/screenshots/en-US/1_ipadPro_1.screenshot-ipad-landscape-1 (1)_framed.png'
[10:14:52]: Framing screenshot './all/screenshots/en-US/1_iphone6Plus_1.screenshot-iphone-1 (2).png'
[10:14:54]: Added frame: '/Users/timothycosta/Desktop/lingq/iLingQ/fastlane/all/screenshots/en-US/1_iphone6Plus_1.screenshot-iphone-1 (2)_framed.png'
[10:14:54]: Framing screenshot './all/screenshots/en-US/2_ipadPro_2.screenshot-ipad-landscape-2 (1).png'
^C+------------------+------------+
|         Lane Context          |
+------------------+------------+
| DEFAULT_PLATFORM | ios        |
| PLATFORM_NAME    | ios        |
| LANE_NAME        | ios frames |
+------------------+------------+

+------+-------------------------------------+-------------+
|                     fastlane summary                     |
+------+-------------------------------------+-------------+
| Step | Action                              | Time (in s) |
+------+-------------------------------------+-------------+
| 1    | Verifying required fastlane version | 0           |
| 2    | default_platform                    | 0           |
| 3    | frameit                             | 8           |
+------+-------------------------------------+-------------+

[10:14:54]: fastlane finished with errors

Cancelled... use --verbose to show the stack trace

##### Output 2

Timothys-Mac-Mini:fastlane timothycosta$ mv all zall
Timothys-Mac-Mini:fastlane timothycosta$ fastlane frames
[10:18:47]: -------------------------------------------------
[10:18:47]: --- Step: Verifying required fastlane version ---
[10:18:47]: -------------------------------------------------
[10:18:47]: Your fastlane version 2.23.0 matches the minimum requirement of 2.20.0  ✅
[10:18:47]: ------------------------------
[10:18:47]: --- Step: default_platform ---
[10:18:47]: ------------------------------
[10:18:47]: Driving the lane 'ios frames' 🚀
[10:18:47]: ---------------------
[10:18:47]: --- Step: frameit ---
[10:18:47]: ---------------------
[10:18:47]: Framing screenshots at path ./fastlane/
[10:18:47]: Using device frames version 'latest'
[10:18:47]: Framing screenshot './background.png'
[10:18:51]: Added frame: '/Users/timothycosta/Desktop/lingq/iLingQ/fastlane/background_framed.png'
[10:18:51]: Framing screenshot './de/screenshots/en-US/1_ipadPro_1.screenshot-ipad-landscape-1 (1).png'
[10:18:57]: Added frame: '/Users/timothycosta/Desktop/lingq/iLingQ/fastlane/de/screenshots/en-US/1_ipadPro_1.screenshot-ipad-landscape-1 (1)_framed.png'
[10:18:57]: Framing screenshot './de/screenshots/en-US/1_iphone6Plus_1.screenshot-iphone-1 (2).png'
[10:18:59]: Added frame: '/Users/timothycosta/Desktop/lingq/iLingQ/fastlane/de/screenshots/en-US/1_iphone6Plus_1.screenshot-iphone-1 (2)_framed.png'
[10:18:59]: Framing screenshot './de/screenshots/en-US/2_ipadPro_2.screenshot-ipad-landscape-2 (1).png'
^C+------------------+------------+
|         Lane Context          |
+------------------+------------+
| DEFAULT_PLATFORM | ios        |
| PLATFORM_NAME    | ios        |
| LANE_NAME        | ios frames |
+------------------+------------+

+------+-------------------------------------+-------------+
|                     fastlane summary                     |
+------+-------------------------------------+-------------+
| Step | Action                              | Time (in s) |
+------+-------------------------------------+-------------+
| 1    | Verifying required fastlane version | 0           |
| 2    | default_platform                    | 0           |
| 3    | frameit                             | 12          |
+------+-------------------------------------+-------------+

[10:19:00]: fastlane finished with errors

Cancelled... use --verbose to show the stack trace

### Environment

Please run `fastlane env` and copy the output below. This will help us help you :+1:
If you used `--capture_output` option please remove this block - as it is already included there.

<details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                       |
| --------------------------- | ------------------------------------------- |
| OS                          | 10.12.3                                     |
| Ruby                        | 2.4.0                                       |
| Bundler?                    | false                                       |
| Git                         | git version 2.10.1 (Apple Git-78)           |
| Installation Source         | /usr/local/bin/fastlane                     |
| Host                        | Mac OS X 10.12.3 (16D32)                    |
| Ruby Lib Dir                | /usr/local/Cellar/ruby/2.4.0/lib            |
| OpenSSL Version             | OpenSSL 1.0.2j  26 Sep 2016                 |
| Is contained                | false                                       |
| Is homebrew                 | false                                       |
| Is installed via Fabric.app | false                                       |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/ |
| Xcode Version               | 8.2.1                                       |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   |             |   |
| LANGUAGE |             |   |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
# Customise this file, documentation can be found here:
# https://github.com/fastlane/fastlane/tree/master/fastlane/docs
# All available actions: https://docs.fastlane.tools/actions
# can also be listed using the `fastlane actions` command

# Change the syntax highlighting to Ruby
# All lines starting with a # are ignored when running `fastlane`

# If you want to automatically update fastlane if a new version is available:
# update_fastlane

# This is the minimum version number required.
# Update this, if you use features of a newer version
fastlane_version "2.20.0"

all_languages = ["es","en","de","ja","ru","fr","sv","it","pt","zh","ko","nl","el"]

default_platform :ios

platform :ios do
  before_all do
    # ENV["SLACK_URL"] = "https://hooks.slack.com/services/..."
    #cocoapods
  end

  desc "Runs all the tests"
  lane :test do
    scan
  end

  desc "Submit a new Beta Build to Apple TestFlight"
  desc "This will also make sure the profile is up to date"
  lane :beta do
    # match(type: "appstore") # more information: https://codesigning.guide
    gym # Build your app - more options available
    pilot

    # sh "your_script.sh"
    # You can also use other beta testing services here (run `fastlane actions`)
  end

  desc "Deploy a new version to the App Store"
  lane :release do
    # match(type: "appstore")
    # snapshot
    gym # Build your app - more options available
    deliver(force: true)
    # frameit
  end

  desc "Upload build, metadata, etc..."
  lane :upload do
    deliver
  end


  desc "Setup provisioning, certificates, capabilities, etc..."
  lane :provision do |options|
    if options[:skip_provision]
        puts "Skipping provisioning"
      next
    end
    produce(
      app_identifier: ENV['APP_IDENTIFIER'],
      app_name: ENV['APP_NAME'],
      language: "English",
      app_version: "1.0",
      sku: ENV['APP_SKU'],
    )
    produce(
      app_identifier: "#{ENV['APP_IDENTIFIER']}.Import-Lesson",
      app_name: "Import Lesson",
      skip_itc: true #Use this to avoid creating a new app and just register the app_identifier
      )
    sh "fastlane produce group -g group.lingq -n 'LingQ Group' --app-identifier #{ENV['APP_IDENTIFIER']}"
    sh "fastlane produce group -g group.lingq -n 'LingQ Group' --app-identifier #{ENV['APP_IDENTIFIER']}.Import-Lesson"
  end

  lane :bootstrap do
    match(type: "appstore")
    match(type: "development")
  end

  desc "A more efficient method of `bootstrap` for all apps because it avoids re-cloning the repository each time"
  lane :bootstrap_all do
    base = "com.lingq.iLingQ"
    apps = [base, "#{base}.Import-Lesson"]
    all_languages.each do |lang|
      apps.push("#{base}.#{lang}")
      apps.push("#{base}.#{lang}.Import-Lesson")
    end
    match(type: "appstore", app_identifier:apps)
    match(type: "development", app_identifier:apps)
  end

  # You can define as many lanes as you want

  after_all do |lane|
    # This block is called, only if the executed lane was successful

    # slack(
    #   message: "Successfully deployed new App Update."
    # )
  end

  error do |lane, exception|
    # slack(
    #   message: exception.message,
    #   success: false
    # )
  end
  desc "Environment Settings"
  lane :check_env do
    puts "APP_IDENTIFIER: #{ENV['APP_IDENTIFIER']}"
    puts "APP_NAME: #{ENV['APP_NAME']}"
    puts "APP_SKU: #{ENV['APP_SKU']}"
    puts "LANG_CODE #{ENV['LANG_CODE']}"
  end

  desc "Building"
  lane :build do |options|
    check_env
    update_project(options)
    provision(options)
    gym
  end

  desc "Setting up xcconfig, plist, etc..."
  lane :update_project do
    fastlane_require 'xcodeproj'
    code = ENV['LANG_CODE']
    path = "../LingQ.xcconfig"
    config = Xcodeproj::Config.new(path)
    config.attributes['LQ_APP_IDENTIFIER'] = ENV['APP_IDENTIFIER']
    config.attributes['LQ_APP_ICON'] = "AppIcon" #code == "all" ? "AppIcon" : "AppIcon-#{code}"
    config.attributes['LQ_APP_NAME'] = ENV['APP_NAME']
    File.write(path, config.to_s) # A bug prevents config from writing itself
    
    # Note that plist is relative to project file
    update_info_plist(
      plist_path: "iLingQ/Info.plist", block: lambda { |plist|
        plist['LQ_LANG_CODE'] = ENV['LANG_CODE']
      }
    )
  end

  desc "Create screenshots"
  lane :screenshots do
    check_env
    update_project
    snapshot
  end

  lane :frames do
    frameit
  end

  desc "Download metadata and screenshots from itunes connect"
  lane :download_itc do
    check_env
    update_project
    sh "fastlane deliver --force download_metadata"
    sh "fastlane deliver --force download_screenshots"
  end

  desc "Create all the folders for each language"
  lane :create_folders do |options|
    fastlane_require 'fileutils'
    puts FileUtils.getwd
    all_languages.each do |lang|
      puts lang
      if options[:delete] && Dir.exist?(lang)
        puts "Deleting #{lang}"
        FileUtils.rm lang, :force => true
      end
      if !Dir.exist?(lang)
        FileUtils.mkdir lang
      end
      FileUtils.cp_r "all/.", lang, :remove_destination => false
      FileUtils.rm_rf "#{lang}/bin/"
    end
  end

  desc "Executes the given lane for all languages plus the main app"
  lane :for_all_apps do |options|
    lane = options[:lane]
    skip = options[:skip_main] ? true : false
    puts "Doing #{lane} for all languages.  Skip main app: #{skip}"
    if !lane
      UI.error "No `lane` parameter passed to `for_all`."
      raise "No `lane` parameter passed to `for_all`."
    end
    if !skip
      sh "fastlane #{lane}"
    end
    all_languages.each do |lang|
      sh "fastlane --env #{lang} #{lane}"
    end
  end

  desc "Copies `files` from all/base_dir/ to LANG_CODE/base_dir/.  Default base_dir is `metadata`"
  lane :copy_files do |options|
    fastlane_require 'fileutils'
    files_string = options[:files]
    base_dir = options[:base_dir] ? options[:base_dir] : "metadata"
    if !files_string
      UI.error "`copy_files requires an argument `files` with a comma separated list, e.g. `fastlane copy_files files:1.txt,2.txt`"
    end
    files = files_string.split(",")
    quit = false
    files.each do |name|
      file = "all/#{base_dir}/#{name}"
      if !File.exist?(file)
        UI.error "No file #{file} exists"
        quit = true
      end
    end
    if quit
      raise "copy_files called with invalid file names"
    end
    files.each do |name|
      file = "all/#{base_dir}/#{name}"
      all_languages.each do |lang|
        copy = "#{lang}/#{base_dir}/#{name}"
        FileUtils.cp file, copy
      end
    end
    sh "fastlane for_all_apps lane:overwrite_name skip_main:1"
  end

  lane :overwrite_name do
    fastlane_require 'fileutils'
    fastlane_require 'pathname'
    lang = ENV['LANG_CODE']
    root = "#{lang}/metadata"
    Dir.entries(root).each do |dir|
      name = "#{root}/#{dir}/name.txt"
      if File.exists?(name)
        File.write(name, ENV['APP_NAME'])
      end
    end
  end

end


# More information about multiple platforms in fastlane: https://github.com/fastlane/fastlane/blob/master/fastlane/docs/Platforms.md
# All available actions: https://docs.fastlane.tools/actions

# fastlane reports which actions are used
# No personal data is recorded. Learn more at https://github.com/fastlane/enhancer
```
</details>

<details><summary>`./fastlane/Appfile`</summary>

```ruby
app_identifier ENV['APP_IDENTIFIER']
apple_id "timothycosta@lingq.com" # Your Apple email address

team_id "FXD5S9TV5H"  # Developer Portal Team ID

# you can even provide different app identifiers, Apple IDs and team names per lane:
# More information: https://github.com/fastlane/fastlane/blob/master/fastlane/docs/Appfile.md
```
</details>

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.23.0  | ✅ Up-To-Date  |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| did_you_mean              | 1.1.0        |
| slack-notifier            | 1.5.1        |
| i18n                      | 0.8.1        |
| thread_safe               | 0.3.6        |
| tzinfo                    | 1.2.2        |
| minitest                  | 5.10.1       |
| activesupport             | 4.2.8        |
| CFPropertyList            | 2.3.5        |
| claide                    | 1.0.1        |
| colored                   | 1.2          |
| nanaimo                   | 0.2.3        |
| xcodeproj                 | 1.4.2        |
| rouge                     | 1.11.1       |
| xcpretty                  | 0.2.4        |
| terminal-notifier         | 1.7.1        |
| unicode-display_width     | 1.1.3        |
| terminal-table            | 1.7.3        |
| plist                     | 3.2.0        |
| public_suffix             | 2.0.5        |
| addressable               | 2.5.0        |
| multipart-post            | 2.0.0        |
| word_wrap                 | 1.0.0        |
| tty-screen                | 0.5.0        |
| babosa                    | 1.0.2        |
| highline                  | 1.7.8        |
| commander-fastlane        | 4.4.4        |
| excon                     | 0.55.0       |
| faraday                   | 0.11.0       |
| unf_ext                   | 0.0.7.2      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20170223 |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| fastimage                 | 2.1.0        |
| gh_inspector              | 1.0.3        |
| uber                      | 0.0.15       |
| representable             | 2.3.0        |
| retriable                 | 2.1.0        |
| mime-types-data           | 3.2016.0521  |
| mime-types                | 3.1          |
| hurley                    | 0.2          |
| little-plugger            | 1.1.4        |
| multi_json                | 1.12.1       |
| logging                   | 2.2.0        |
| jwt                       | 1.5.6        |
| memoist                   | 0.15.0       |
| os                        | 0.9.6        |
| signet                    | 0.7.3        |
| googleauth                | 0.5.1        |
| httpclient                | 2.8.3        |
| google-api-client         | 0.9.28       |
| json                      | 2.0.2        |
| mini_magick               | 4.5.1        |
| multi_xml                 | 0.6.0        |
| rubyzip                   | 1.2.1        |
| security                  | 0.1.3        |
| xcpretty-travis-formatter | 0.0.4        |
| dotenv                    | 2.2.0        |
| bundler                   | 1.14.6       |
| faraday_middleware        | 0.11.0.1     |
| io-console                | 0.4.6        |
| openssl                   | 2.0.2        |
</details>

*generated on:* **2017-03-28**
</details>

**Comment by timothycosta** [comment]
Well, I was able to at least handle question 1 with the following:

```
  lane :frames do
    Dir.chdir ENV['APP_DIR'] do
      sh "fastlane frameit"
    end
  end
```

Which brings up general point about documentation.  As I've been working on integrating fastlane, it's been difficult to figure out how all the "configuration bits" mesh together and what should be called via `sh` vs programmatically.  Some functionality appears to only be available via one or the other, some via both (take a look at the `provision` lane in my Fastfile above.)  And in each case, it's not perfectly clear where the command will take its configuration from, how the current directory might affect things, if there's a difference between `frameit` or `sh "fastlane frameit"` (in `:frames` above, there was).

Admittedly, this might all be more clear to experienced ruby developers compared to iOS/mac devs.  

(Above meant constructively, not critically! :))

**Comment by KrauseFx** [comment]
Hey @timothycosta, sorry for the late reply. Unfortunately we haven't worked on _frameit_ recently, and it lacks of a good documentation. What are the open questions you still have?

**Comment by timothycosta** [comment]
Thanks @KrauseFx .  I ended up rolling my own tool here, but answering these may help other users.  Maybe the easiest thing is to copy my questions and reorganize them a bit.

### Questions:
1. [x] Is there currently any way to get frameit to work on one `[app, directory, .env]` at a time? (Yes, see my comment above)
2. [ ] Frameit doesn't appear to delete the unframed files.  How does `deliver` determine which files are the correct ones?
3. [ ] Can a file be excluded from framing? -- I would think yes if it doesn't match a `filter`, but strangely I was ending up with `background.png` getting framed at one point...
2. [ ] How does Framefile.json work?
    * [x] For example, it's not clear how the filters apply.  Do they get used only once, or for every matching occurrence of that string in the filename?  Probably the latter, but being explicit is nice.  (It's clear now `filter` is applied to *any* matching image in a given directory)
    * [ ] What does `padding` do?  Is it horizontal, vertical?  Above and below the text or just below?
3. [ ] Documentation should be provided regarding the `background` file.  
    * [ ] It appears this is used for all images, but there is no suggestion on size, or mention of how it gets scaled.  (Relevant if you wanted to use a gradient image.)


### Suggestions:
1. [ ] The current behavior should be added to the documentation regarding how frameit determines which files/directories to work on/in. (Maybe update documentation with my fix above, or something better.)
1. [ ] Allow for a separate `Framefile` to match other fastlane tools.  Could include something like `output_directory` which matches `Snapfile`, `keywords_directory`, `titles_directory`, `fonts_directory`, `json_path`...
3. Background File:
    * [ ] Allow hex value in .json file instead of an image

### General Documentation
To sort of summarize my previous comment, it's confusing how there are actions and methods with the same name, and where all of those actually take their values from.

I ended up learning a lot more ruby than I expected when I started out with Fastlane, so a "Ruby for iOS Developers using Fastlane" tutorial might go a long way.  I'll try to illustrate with an example from my Fastfile:

```
    produce(
      app_identifier: "#{ENV['APP_IDENTIFIER']}.Import-Lesson",
      app_name: "Import Lesson",
      skip_itc: true #Use this to avoid creating a new app and just register the app_identifier
      )
    sh "fastlane produce group -g group.lingq -n 'LingQ Group' --app-identifier #{ENV['APP_IDENTIFIER']}"
```

Not being a proper ruby dev, it's not clear to me whether I'm calling the action `produce` or the method, or if there's a difference, and in looking at documentation/issues it's often hard to grok which one is being referred to because of ruby's flexible syntax.  But as an objc/swift developer what I'd really like to do in my Fastfile is like (pseudocode) `Fastlane.Produce.create_group(inject_all_my_dependencies...)` `Fastlane.Produce.register(inject_all_my_dependencies...)`.  

Then in the documentation, there would be `Fastlane.Produce.create_group - Equivalent to "fastlane produce create_group" - arg1: defaults to Appfile.KEY, or ENV['KEY']...`

The opposite would be just as good - `fastlane action produce ...` could list out how to call the method programmatically.
